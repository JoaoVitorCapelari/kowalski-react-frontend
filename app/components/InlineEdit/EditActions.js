import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import FaCheck from 'react-icons/lib/fa/check';
import TiTimes from 'react-icons/lib/ti/times';

const EditActionsWrapper = styled.div`
  display: flex;
  flex-grow: 0;
  position: absolute;
  bottom: -28px;
  right: 0px;
`;

const CheckChangesButton = styled.a`
  display: flex;
  color: inherit;
  margin-left: 2px;
  margin-right: 1px;
`;

const DiscardChangesButton = styled.a`
  display: flex;
  color: inherit;
  margin-left: 1px;
  margin-right: 2px;
`;

function EditActions({ onConfirmChanges, onDiscardChanges }) {
  return (
    <EditActionsWrapper>
      <CheckChangesButton title="Confirm changes" onMouseDown={onConfirmChanges}>
        <FaCheck size={19} />
      </CheckChangesButton>
      <DiscardChangesButton title="Discard changes" onMouseDown={onDiscardChanges}>
        <TiTimes size={22} />
      </DiscardChangesButton>
    </EditActionsWrapper>
  );
}

EditActions.propTypes = {
  onConfirmChanges: PropTypes.func.isRequired,
  onDiscardChanges: PropTypes.func.isRequired,
};

export default EditActions;
