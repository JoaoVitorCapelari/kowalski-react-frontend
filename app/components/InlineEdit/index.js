/**
*
* InlineEdit
*
*/

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import FaPencil from 'react-icons/lib/fa/pencil';
import FaCheck from 'react-icons/lib/fa/check';
import TiTimes from 'react-icons/lib/ti/times';
import { ClipLoader } from 'react-spinners';

const Container = styled.div`
  display: flex;
  flex-grow: 1;

  ${({ stateName }) => (
    [STATE_HOVERED, STATE_EDITING].includes(stateName) ? `
      outline: 1px solid grey;
    ` : ''
  )}
`;

const ContentContainer = styled.div`
  display: flex;
  flex-grow: 1;
  position: relative;
  width: 100%;
`;

const ButtonsContainer = styled.a`
  display: none;
  color: inherit;
  margin-right: 5px;
  flex-grow: 0;

  ${({ stateName }) => (
    [STATE_HOVERED, STATE_SAVING].includes(stateName) ? `
      display: block;
    ` : ''
  )}
`;

const EditActionsWrapper = styled.div`
  display: flex;
  flex-grow: 0;
  position: absolute;
  bottom: -28px;
  right: 0px;
`;

const EditButton = styled.div`
`;

const CheckChangesButton = styled.a`
  display: flex;
  color: inherit;
  margin-left: 2px;
  margin-right: 1px;
`;

const DiscardChangesButton = styled.a`
  display: flex;
  color: inherit;
  margin-left: 1px;
  margin-right: 2px;
`;

export const STATE_NORMAL = 'STATE_NORMAL';
export const STATE_HOVERED = 'STATE_HOVERED';
export const STATE_EDITING = 'STATE_EDITING';
export const STATE_SAVING = 'STATE_SAVING';

const TextBeingSaved = styled.span`
  font-style: italic;
  color: grey;
`;

// TODO Refactor to be a presentational component, should not have internal state
// InlineEdit =>
//  Presentational component
//  Props:
//    - state(saving, normal, hovered, editing)
//    - value
//    - onCommit onDiscard onChange
//    - children: Funcao para renderizacao do componente de edicao ()
// InlineEdit =>
//  Presentational component
//  Props:
class InlineEdit extends Component {
  constructor(props) {
    super(props);
    this.state = { name: props.saving ? STATE_SAVING : STATE_NORMAL, content: props.value };

    this.unhover = this.unhover.bind(this);
    this.hover = this.hover.bind(this);
    this.edit = this.edit.bind(this);
    this.save = this.save.bind(this);
    this.contentChanged = this.contentChanged.bind(this);
    this.dicardChanges = this.dicardChanges.bind(this);
    this.commitChanges = this.commitChanges.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    // Whenever a new value is received, update the content
    if (this.props.value !== nextProps.value) {
      this.setState({
        ...this.state,
        content: nextProps.value,
      });
    }

    if (!!nextProps.saving !== (this.state.name === STATE_SAVING)) {
      this.setState({
        ...this.state,
        name: nextProps.saving ? STATE_SAVING : STATE_NORMAL,
        content: nextProps.saving ? this.state.content : this.props.value,
      });
    }
  }

  contentChanged(content) {
    this.setState({ ...this.state, content });
  }

  unhover() {
    if (this.state.name === STATE_HOVERED) {
      this.setState({ ...this.state, name: STATE_NORMAL });
    }
  }

  dicardChanges() {
    this.setState({ name: STATE_NORMAL, content: this.props.value });
  }

  commitChanges() {
    this.props.onCommit(this.state.content);
  }

  hover() {
    if (this.state.name === STATE_NORMAL) {
      this.setState({ ...this.state, name: STATE_HOVERED });
    }
  }

  edit() { this.setState({ ...this.state, name: STATE_EDITING }); }
  save() { this.setState({ ...this.state, name: STATE_SAVING }); }

  formatValue() {
    return this.props.formatter ? this.props.formatter(this.props.value) : this.props.value;
  }

  renderButtonsContainerForEditingState(stateName) {
    return (
      <EditActionsWrapper stateName={stateName}>
        <CheckChangesButton title="Confirm changes" onMouseDown={this.commitChanges}>
          <FaCheck size={19} />
        </CheckChangesButton>
        <DiscardChangesButton title="Discard changes" onMouseDown={this.dicardChanges}>
          <TiTimes size={22} />
        </DiscardChangesButton>
      </EditActionsWrapper>
    );
  }

  renderEditComponent() {
    return this.props.renderEditComponent({
      content: this.state.content,
      onCommit: this.commitChanges,
      onDiscard: this.dicardChanges,
      onChange: this.contentChanged,
      stateName: this.state.name,
    });
  }

  renderContentOnSaving() {
    if (this.props.displayEditElementWhenSaving) {
      return <TextBeingSaved>{this.renderEditComponent()}</TextBeingSaved>;
    }

    return (
      <TextBeingSaved>
        {
          this.props.formatter
            ? this.props.formatter(this.state.content)
            : this.state.content
        }
      </TextBeingSaved>
    );
  }

  renderButtonsContainerNonEditingState() {
    return (
      <EditButton>
        { this.state.name !== STATE_SAVING && <FaPencil size={18} /> }
        { this.state.name === STATE_SAVING && <ClipLoader size={18} /> }
      </EditButton>
    );
  }

  renderContentOnHovered() {
    const formattedValue = this.formatValue();
    if (this.props.displayEditElementWhenOnHover) {
      return this.renderEditComponent();
    }

    return formattedValue;
  }

  renderContentOnNormalState() {
    const formattedValue = this.formatValue();
    if (this.props.displayEditElementWhenNormal) {
      return this.renderEditComponent();
    }

    return formattedValue;
  }

  render() {
    return (
      <Container
        onMouseEnter={this.hover}
        onMouseLeave={this.unhover}
        stateName={this.state.name}
        onClick={this.edit}
        title={'Click for edit'}
      >
        <ContentContainer>
          { this.state.name === STATE_NORMAL && this.renderContentOnNormalState() }
          { this.state.name === STATE_HOVERED && this.renderContentOnHovered() }
          { this.state.name === STATE_EDITING && this.renderEditComponent() }
          { this.state.name === STATE_EDITING && this.renderButtonsContainerForEditingState(this.state.name) }
          { this.state.name === STATE_SAVING && this.renderContentOnSaving() }
        </ContentContainer>
        <ButtonsContainer stateName={this.state.name}>
          { this.state.name !== STATE_EDITING && this.renderButtonsContainerNonEditingState() }
        </ButtonsContainer>
      </Container>
    );
  }
}

InlineEdit.propTypes = {
  formatter: PropTypes.func, // Optional function to format the value shown
  onCommit: PropTypes.func.isRequired,
  saving: PropTypes.bool,
  renderEditComponent: PropTypes.func.isRequired,
  displayEditElementWhenSaving: PropTypes.bool,
  displayEditElementWhenOnHover: PropTypes.bool,
  displayEditElementWhenNormal: PropTypes.bool,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.instanceOf(Date)]),
};

export default InlineEdit;
