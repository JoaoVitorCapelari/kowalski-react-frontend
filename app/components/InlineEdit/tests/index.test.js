import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

import InlineEdit, { STATE_EDITING, STATE_SAVING } from '../index';

const renderComponent = (props = {}) => shallow(
  <InlineEdit onCommit={jest.fn()} renderEditComponent={jest.fn()} {...props} />
);

describe('<InlineEdit />', () => {
  describe('when in normal state', () => {
    it('renders', () => {
      const renderedComponent = renderComponent();
      expect(renderedComponent.html()).toMatchSnapshot();
    });

    it('becomes outlined on mouseEnter', () => {
      const renderedComponent = renderComponent();
      expect(renderedComponent).not.toHaveStyleRule('outline', /solid/);
      renderedComponent.simulate('mouseEnter');
      expect(renderedComponent).toHaveStyleRule('outline', /solid/);
    });
  });

  describe('when is hovered', () => {
    let renderedComponent;

    beforeEach(() => {
      renderedComponent = renderComponent();
      renderedComponent.simulate('mouseEnter');
    });

    it('renders outlined', () => {
      expect(renderedComponent).toHaveStyleRule('outline', /solid/);
      expect(renderedComponent.html()).toMatchSnapshot();
    });

    it('goes to editing state on click', () => {
      renderedComponent.simulate('click');
    });

    it('goes to normal state on mouseLeave', () => {
      expect(renderedComponent).toHaveStyleRule('outline', /solid/);
      expect(renderedComponent.html()).toMatchSnapshot();
    });

    it('removes outline when mouse leaves', () => {
      const renderedComponent = renderComponent();
      renderedComponent.simulate('mouseEnter');
      expect(renderedComponent).toHaveStyleRule('outline', /solid/);
      renderedComponent.simulate('mouseLeave');
      expect(renderedComponent).not.toHaveStyleRule('outline', /solid/);
    });
  });

  describe('when editing content', () => {
    it('renders itself outlined and calls renderEditComponent', () => {
      const renderEditComponent = jest.fn();
      const renderedComponent = renderComponent({ value: 'one', renderEditComponent });
      renderedComponent.simulate('click');
      expect(renderedComponent).toHaveStyleRule('outline', /solid/);
      expect(renderEditComponent.mock.calls.length).toBe(1);
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('content', 'one');
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('onCommit');
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('onDiscard');
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('onChange');
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('stateName', STATE_EDITING);
      expect(renderedComponent.html()).toMatchSnapshot();
    });
  });

  describe('when saving content', () => {
    it('renders', () => {
      const renderEditComponent = jest.fn();
      const renderedComponent = renderComponent({ value: 'one', displayEditElementWhenSaving: true, renderEditComponent });
      renderedComponent.simulate('click');
      expect(renderEditComponent.mock.calls.length).toBe(1);
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('stateName', STATE_EDITING);
      renderEditComponent.mock.calls[0][0].onChange('two');
      renderEditComponent.mock.calls[0][0].onCommit();
      renderedComponent.setProps({ saving: true });
      expect(renderedComponent).not.toHaveStyleRule('outline', /solid/);
      expect(renderEditComponent.mock.calls.length).toBe(2);
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('content', 'two');
      expect(renderEditComponent.mock.calls[0][0]).toHaveProperty('stateName', STATE_SAVING);
    })
  });


  describe('when receiving new value', () => {
    it('updates the content accordingly', () => {
      const renderedComponent = renderComponent({ value: 'one' });
      expect(renderedComponent.state().content).toBe('one');
      renderedComponent.setProps({ value: 'newValue' });
      expect(renderedComponent.state().content).toBe('newValue');
    });
  });

  // describe('when receiving new saving status', () => {
  //   it('updates the state accordingly', () => {
  //     const renderedComponent = renderComponent({ value: 'one' });
  //     expect(renderedComponent.state().content).toBe('one');
  //     renderedComponent.setProps({ value: 'newValue' });
  //     expect(renderedComponent.state().content).toBe('newValue');
  //     renderedComponent.setProps({ saving: true });
  //     expect(renderedComponent.state().name).toBe(STATE_SAVING);
  //   });
  // });

  // describe('contentChanged', () => {
  //   it('updates the state accordingly', () => {
  //     const renderedComponent = renderComponent({ value: 'one' });
  //     expect(renderedComponent.state().content).toBe('one');
  //     renderedComponent.instance().contentChanged('new');
  //     expect(renderedComponent.state().content).toBe('new');
  //   });
  // });

  // describe('hover', () => {
  //   describe('when state is already hovered', () => {
  //     it('updates the state accordingly', () => {

  //     });
  //   });

  //   describe('when state is not hovered', () => {

  //   });

  //   it('updates the state accordingly', () => {
  //     const renderedComponent = renderComponent({ value: 'one' });
  //     expect(renderedComponent.state().content).toBe('one');
  //     renderedComponent.instance().contentChanged('new');
  //     expect(renderedComponent.state().content).toBe('new');
  //   });
  // });
});
