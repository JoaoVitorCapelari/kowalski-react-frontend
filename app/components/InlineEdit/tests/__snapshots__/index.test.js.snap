// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineEdit /> when editing content renders 1`] = `"<div class=\\"InlineEdit__Container-iwjTUk imIxCw\\" title=\\"Click for edit\\"><div class=\\"InlineEdit__ContentContainer-kvxWLC cdZpQT\\"><div class=\\"InlineEdit__EditActionsWrapper-eykUov hpOFXo\\"><a class=\\"InlineEdit__CheckChangesButton-fRhjES cHvrUT\\" title=\\"Confirm changes\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"19\\" width=\\"19\\" viewBox=\\"0 0 40 40\\" style=\\"vertical-align:middle;\\"><g><path d=\\"m37.3 12.6q0 0.9-0.6 1.6l-19.2 19.1q-0.6 0.7-1.5 0.7t-1.6-0.7l-11.1-11.1q-0.6-0.6-0.6-1.5t0.6-1.5l3.1-3q0.6-0.7 1.5-0.7t1.5 0.7l6.6 6.5 14.6-14.6q0.6-0.6 1.5-0.6t1.5 0.6l3.1 3q0.6 0.6 0.6 1.5z\\"></path></g></svg></a><a class=\\"InlineEdit__DiscardChangesButton-gqFUEn bWYrP\\" title=\\"Discard changes\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"22\\" width=\\"22\\" viewBox=\\"0 0 40 40\\" style=\\"vertical-align:middle;\\"><g><path d=\\"m29 11c-1.3-1.3-3.4-1.3-4.7 0l-4.3 4.3-4.3-4.3c-1.3-1.3-3.4-1.3-4.7 0-1.3 1.3-1.3 3.4 0 4.7l4.3 4.3-4.3 4.3c-1.3 1.3-1.3 3.4 0 4.7 0.6 0.7 1.5 1 2.3 1s1.7-0.3 2.4-1l4.3-4.3 4.3 4.3c0.7 0.7 1.5 1 2.4 1s1.7-0.3 2.3-1c1.3-1.3 1.3-3.4 0-4.7l-4.3-4.3 4.3-4.3c1.3-1.3 1.3-3.4 0-4.7z\\"></path></g></svg></a></div></div><a class=\\"InlineEdit__ButtonsContainer-cWlMqY hnRZzJ\\"></a></div>"`;

exports[`<InlineEdit /> when in normal state renders 1`] = `"<div class=\\"InlineEdit__Container-iwjTUk hrrFYE\\" title=\\"Click for edit\\"><div class=\\"InlineEdit__ContentContainer-kvxWLC cdZpQT\\"></div><a class=\\"InlineEdit__ButtonsContainer-cWlMqY hnRZzJ\\"><div class=\\"InlineEdit__EditButton-cDVLLs vgvHz\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"18\\" width=\\"18\\" viewBox=\\"0 0 40 40\\" style=\\"vertical-align:middle;\\"><g><path d=\\"m11.1 34.3l2-2-5.2-5.3-2 2v2.4h2.8v2.9h2.4z m11.7-20.7q0-0.5-0.5-0.5-0.2 0-0.4 0.1l-12.1 12.1q-0.1 0.2-0.1 0.4 0 0.5 0.4 0.5 0.3 0 0.4-0.2l12.1-12.1q0.2-0.1 0.2-0.3z m-1.2-4.3l9.3 9.3-18.6 18.5h-9.3v-9.2z m15.2 2.1q0 1.2-0.8 2l-3.7 3.7-9.3-9.2 3.7-3.7q0.8-0.9 2-0.9 1.2 0 2 0.9l5.3 5.2q0.8 0.9 0.8 2z\\"></path></g></svg></div></a></div>"`;

exports[`<InlineEdit /> when is hovered renders 1`] = `"<div class=\\"InlineEdit__Container-iwjTUk imIxCw\\" title=\\"Click for edit\\"><div class=\\"InlineEdit__ContentContainer-kvxWLC cdZpQT\\"></div><a class=\\"InlineEdit__ButtonsContainer-cWlMqY dYxLvb\\"><div class=\\"InlineEdit__EditButton-cDVLLs vgvHz\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"18\\" width=\\"18\\" viewBox=\\"0 0 40 40\\" style=\\"vertical-align:middle;\\"><g><path d=\\"m11.1 34.3l2-2-5.2-5.3-2 2v2.4h2.8v2.9h2.4z m11.7-20.7q0-0.5-0.5-0.5-0.2 0-0.4 0.1l-12.1 12.1q-0.1 0.2-0.1 0.4 0 0.5 0.4 0.5 0.3 0 0.4-0.2l12.1-12.1q0.2-0.1 0.2-0.3z m-1.2-4.3l9.3 9.3-18.6 18.5h-9.3v-9.2z m15.2 2.1q0 1.2-0.8 2l-3.7 3.7-9.3-9.2 3.7-3.7q0.8-0.9 2-0.9 1.2 0 2 0.9l5.3 5.2q0.8 0.9 0.8 2z\\"></path></g></svg></div></a></div>"`;
